dbt_meshify/utilities/grouper.py

```diff
@@ -0,0 +1,131 @@
+import os
+from pathlib import Path
+from typing import Optional, Dict, Set, Tuple
+
+import networkx
+from dbt.contracts.graph.nodes import Group, ModelNode
+from dbt.contracts.graph.unparsed import Owner
+from dbt.node_types import AccessType, NodeType
+
+from dbt_meshify.dbt_projects import DbtProject
+from dbt_meshify.storage.file_manager import DbtFileManager
+from dbt_meshify.storage.yaml_editors import DbtMeshYmlEditor
+
+
+class ResourceGroupingException(BaseException):
+    """Exceptions relating to grouping of resources."""
+
+
+class ResourceGrouper:
+    """
+    The ResourceGrouper is responsible for generating a dbt-core Group for a
+    collection of resources in a DbtProject, and for providing access control
+    recommendations based on the reference characteristics for each resource.
+    """
+
+    def __init__(self, project: DbtProject):
+        self.project = project
+        self.meshify = DbtMeshYmlEditor()
+        self.file_manager = DbtFileManager(read_project_path=project.path)
+
+    @classmethod
+    def identify_interface(cls, graph: networkx.Graph, selected_bunch: Set[str]) -> Set[str]:
+        """
+        Given a graph, find the interface nodes, where interface nodes are the nodes that
+        either have downstream consumers outside the selected bunch OR are nodes
+        with no downstream consumers.
+        """
+        boundary_nodes = {edge[0] for edge in networkx.edge_boundary(graph, selected_bunch)}
+        leaf_nodes = {node for node, out_degree in graph.out_degree() if out_degree == 0}
+        return boundary_nodes | leaf_nodes
+
+    @classmethod
+    def classify_resource_access(cls, graph, nodes):
+        """
+        Identify what access types each node should have. We can make some simplifying assumptions about
+        recommended access for a group.
+
+        For example, interfaces (nodes on the boundary of a subgraph or leaf nodes) should be public,
+        whereas nodes that are not a referenced are safe for a private access level.
+        """
+        boundary_nodes = cls.identify_interface(graph, nodes)
+        print(boundary_nodes)
+        resources = {
+            node: AccessType.Public if node in boundary_nodes else AccessType.Private
+            for node in nodes
+        }
+        return resources
+
+    def _generate_resource_group(
+        self,
+        name: str,
+        owner: Owner,
+        path: os.PathLike,
+        select: str,
+        exclude: Optional[str] = None,
+    ) -> Tuple[Group, Dict[str, AccessType]]:
+        """Generate the ResourceGroup that we want to apply to the project."""
+
+        group = Group(
+            name=name,
+            owner=owner,
+            package_name=self.project.name,
+            original_file_path=os.path.relpath(path, self.project.path),
+            # TODO: This seems a bit yucky. How does dbt-core generate unique_ids?
```

Comment:
[appears to be just how you did it](https://github.com/dbt-labs/dbt-core/blob/8f998c218eb8ec37bc3db99ac93148be755f332c/core/dbt/parser/base.py#L52)

Reply:
Lol! I guess we're good to go, then 🤣 

```suggestion
```